### DISCUSSION TITLE
SvelteKit Auth - A Nightmare

### DISCUSSION THREAD

**Discussion by ethanfox** [supabase/supabase#13835] [Questions] [answered]
Been trying to implement auth on Sveltekit with multiple failed attempts.

I am now going through the _Build a User Management App with SvelteKit_ **again** and I am getting the following error:
`TypeError: getSession is not a function
    at load (+layout.server.ts:6:19)
    at Module.load_server_data (/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:43:41)
    at /node_modules/@sveltejs/kit/src/runtime/server/page/index.js:141:18
    at processTicksAndRejections (node:internal/process/task_queues:96:5)`

Here's the code:

```
// src/routes/+layout.server.ts
import type { LayoutServerLoad } from "./$types";

export const load: LayoutServerLoad = async ({ locals: { getSession } }) => {
  return {
    session: await getSession(),
  };
};

```

**Comment by silentworks** [comment]
Can you share your `+hooks.server.ts` file here? as this is where the `getSession` is setup before it can be accessed inside of a `+layout.server.ts` file. Also share what version of SvelteKit you are using in this instance.

_This is Andrew who you've been communicating with thru support._

**Reply by ethanfox** [reply]
```
// src/hooks.server.ts
import {
  PUBLIC_SUPABASE_URL,
  PUBLIC_SUPABASE_ANON_KEY,
} from "$env/static/public";
import { createSupabaseServerClient } from "@supabase/auth-helpers-sveltekit";
import type { Handle } from "@sveltejs/kit";

export const handle: Handle = async ({ event, resolve }) => {
  event.locals.supabase = createSupabaseServerClient({
    supabaseUrl: PUBLIC_SUPABASE_URL,
    supabaseKey: PUBLIC_SUPABASE_ANON_KEY,
    event,
  });

  /**
   * A convenience helper so we can just call await getSession() instead const { data: { session } } = await supabase.auth.getSession()
   */
  event.locals.getSession = async () => {
    const {
      data: { session },
    } = await event.locals.supabase.auth.getSession();
    return session;
  };

  return resolve(event, {
    /**
     * There´s an issue with `filterSerializedResponseHeaders` not working when using `sequence`
     *
     * https://github.com/sveltejs/kit/issues/8061
     */
    filterSerializedResponseHeaders(name) {
      return name === "content-range";
    },
  });
};

```

**Reply by silentworks** [reply]
Can you create a test repo with all the files you have currently in your project on GitHub and post a link here because this code looks correct so not sure why you would be getting an error. Also have you tried any of the example projects I linked you to?

**Reply by silentworks** [reply]
@ethanfox also please use a more appropriate title for this post as **SvelteKit Auth - A Nightmare** doesn't help anyone having the same issues as you. The more descriptive the title the better for others who help out on here too.

**Reply by ethanfox** [reply]
Just FYI I started this new one from scratch to specifically test authentication(Created anew project and all of that). There's too much going on in my actual project to be changing this stuff. 

No I cant get them to work

**Reply by silentworks** [reply]
What failed in them specifically?

**Reply by ethanfox** [reply]
Now they're working. Don't know what changed.

**Reply by silentworks** [reply]
Can you see anything that is different in the example projects and your current code?

**Reply by silentworks** [reply]
@ethanfox I've had a look at your repo and in the root your layout.ts file is missing the `+`. It should be `+layout.ts` this is the reason why `supabase` is undefined because this file handles passing the supabase instance over to the `.svelte` pages.

**Reply by silentworks** [reply]
@ethanfox please continue in this thread, I've outlined above one of the issue in your project and I've also replied to you on support email about the other issues.

**Reply by ethanfox** [reply]
Readingover the migration guide - I now have to rewrite every function that calls to the database?  Still getting an error
TypeError: Cannot read properties of undefined (reading 'session')
```
    let loadedData = [];
    async function loadData() {
        const { data: result } = await data.supabase.from('test').select('*').limit(20);
        loadedData = result;
    }

    $: if (data.session) {
        //loadData();
    }
    onMount(() => {
        fetchUserData();
    });```

**Reply by silentworks** [reply]
Have you fixed the issues I've mentioned in the above to do with your `layout.ts` file missing the `+`?

**Reply by ethanfox** [reply]
yes

**Reply by silentworks** [reply]
In the example you paste above I can't see where `data` is defined, can you share that part of the code please or tell me where in your project I would find this file and also can you push your latest changes up to your project repo which you've shared with me.

**Reply by ethanfox** [reply]
pushed

**Reply by silentworks** [reply]
I've just looked at the code, I think you will need to spend some time in the repo as you are doing some stuff which are not even SvelteKit standard way of doing things. The error you are getting now is because you are trying to load PageData into a component, this is not something you can do directly, PageData is only available to `+page.svelte` files inside of routes not in components. You will need to pass the data you need from the `+page.svelte` to that component(`CheckUser.svelte` in this case). This should be the same throughout every component in your app that wants to make use of the supabase instance from the auth-helpers-sveltekit.

**Reply by ethanfox** [reply, marked as answer]
I'll just rebuild the app in Firebase. Well thought-out documentation would have prevented this. ✌️

**Reply by silentworks** [reply]
I wish you all the best with your app and hopefully you will try us again in the future.

**Reply by anasmohammed361** [reply]
Firebase is a really awesome tool , i often find myself not liking a nosql database for any major projects as it just causes problems or ineffeciency while querying (my pov) , You should really look into your  problems before just jumping around with tech stacks.

**Comment by ethanfox** [comment]


**Reply by silentworks** [reply]
I've followed the user management tutorial step by step and didn't hit any of these issues. I think unless you create a repo with your code it will be hard to say exactly what is going wrong. The examples I sent you a link to are all based off the same code as the user management tutorial and those seem to be working too so not sure what's going on in your case.

From the code above it's saying that `supabase` is undefined so it can't read `auth` off of it. The issue is elsewhere in your code.

**Comment by ethanfox** [comment]


**Reply by silentworks** [reply]
Please try and keep the conversation in one thread. I've outlined some of the issues in the support email and the main one above in the other thread https://github.com/orgs/supabase/discussions/13835#discussioncomment-5666178

**Reply by youfun** [reply]
###  i have the same problem . i ues the code from https://supabase.com/docs/guides/auth/auth-helpers/sveltekit
I did not change a single word

**Comment by ethanfox** [comment]
> ### i have the same problem . i ues the code from https://supabase.com/docs/guides/auth/auth-helpers/sveltekit
> I did not change a single word


In order to get this to work I ended up following Huntabyte's tutorial that goes over using Sveltekit Stripe and Suapabse. It works flawlessly for me now but did not use what the Supaabse docs have. It costs like $130 which sucks but it covers a lot and the videos are quick and not drawn out like a lot of other tutorials.

**Comment by umutahmet** [comment]
For anyone else facing this issue, make sure your file is named correctly. @silentworks it should be named `hooks.server.{js|ts}` not `+hooks.server.{js|ts}` (no plus). My issue was another stupid typo in the filename 😅 but this isn't about me haha. This thread helped me overcome my stupidity. Example works for me fine 💪 Thanks!

**Reply by nexttrack07** [reply]
Actually my mistake was similar, I put the hooks.server.ts file inside the routes/ folder instead of the src/ folder. **The correct place is src/hooks.server.ts**

**Comment by internpoon** [comment]
@silentworks 

If i have 2 tabs opened, I signed out on 1 tab, I am still able to browse the protected route on the other tab. 

I used a POST API to signout and in the .svelte file i used a goto('/') to redirect to the login page, not sure if this is causing it but i doubt.

**Reply by anasmohammed361** [reply]
Yes it is a known problem with sveltekit  , please do know that it has nothing to do with supabase there are two known workarounds for it. https://www.youtube.com/watch?v=UbhhJWV3bmI , https://www.youtube.com/watch?v=K1Tya6ovVOI

https://github.com/sveltejs/kit/issues/6315

**Reply by internpoon** [reply]
I tried to do with Firebase and I dont have this issue. And yes, I used server hook for a small workaround, but please note that if any of your route without `+page.server.ts` would not trigger the hook and still bypassable I guess.

Add below to your hook to make it work for any one who is looking for a workaround. Also, every route make sure to have a `+page.server.ts`
```
  let protectedRoute = ['routeA', 'routeB', 'routeC', 'routeD']
  if (protectedRoute.includes(event.url.pathname.substring(1))) {
    const session = await event.locals.getSession()
    if (!session) {
      throw redirect(303, '/')
    }
  }
  ```

**Reply by anasmohammed361** [reply]
Yes `hooks.server.ts` does'nt run for csr pages, By default Firebase creds are stored in local storage and it uses Client side for authenticating users. If you are using `firebase-admin` and using cookies for authentication , i think it will behave the same.

**Comment by codepilotsf** [comment]
With SK 2.0 now the default, the current docs are completely broken. NPM can't even install deps without `--force`.  And even then I wasn't able to get it working.

**Comment by ybaamel** [comment]
**@silentworks Is there any updates on the docs to solve this problem?**

**Comment by codepilotsf** [comment]
A very friendly guy named Francis that I met on Discord made this starter with a helpful README and it seems to be exactly what we were looking for: https://github.com/fnimick/sveltekit-supabase-auth-starter

Hope that helps!

**Comment by charislam** [comment]
Thanks @codepilotsf for sharing the resource! Hope others will find it useful. One of our priorities is revisiting our Auth SSR documentation overall, which we agree could use some improvements. We'll start with Next.js, but SvelteKit is most likely next on the list.

We really appreciate the community jumping in to share resources and help in the meantime!

**Comment by astrixlabs** [comment]
How is this issue still not fixed? All it requires is a documentation or tutorial update. This is disappointing.

**Reply by anasmohammed361** [reply]
i'll drop a sample repo if you can share you can give me a description of the issue.

**Comment by charislam** [comment]
@astrixlabs do you have a recent repro where the SvelteKit instructions don't work? There were some tweaks lately to fix the SvelteKit instructions (making them compatible with the cookies methods on SvelteKit 2.0)

**Reply by astrixlabs** [reply]
I got it working, it's fine thanks

**Comment by psegarel** [comment]
[Supabase Auth / SvelteKit 4 Starter ( Typescript )](https://github.com/psegarel/supabase_auth-sveltekit4)
Repo for a starter template. Slightly different than the one posted above as it uses token_hash & type as opposed to code in the api endpoint

**Comment by prymorsk** [comment]
I want to love supabase + sveltekit but I am struggling as well. My issues are not just getting Auth working, I've got that part going thanks to moving to @ssr package and following this tutorial [https://dev.to/kvetoslavnovak/supabase-ssr-auth-48j4].   My issue is trying to use all this with Vercel's AI SDK. Making everything work together is too much for me. Just running an authenticated query from /api/chat is a complete mystery to me.

**Reply by J0** [reply]
Hey @prymorsk,

Sorry to hear about the issues - mind sharing more about the errors you're facing? We'd love to try and be helpful if possible.

Let us know!

**Reply by chufucious** [reply]
Yeah. Can't get it to work with SvelteKit.

**Comment by Anderneku** [comment]
I actually had this issue as well but what worked for me was moving my hooks.server.ts from src/hooks to src/

**Comment by tyforcode** [comment]
https://github.com/fnimick/sveltekit-supabase-auth-starter was extremely helpful for me and got things working the way you'd expect. The docs need to be updated
